rfd/0030-client-side-session-encryption.md

```diff
@@ -0,0 +1,168 @@
+
+The envelope header (see "envelope header", below) is treated as part of the payload, and included in the payload byte count.
+
+The `Flags` field is left open for future expansion.
+
+##### Defined Flags
+|                  Flag |      Name | Interpretation                              |
+|-----------------------|-----------|---------------------------------------------|
+| `0x00000000000000001` | Encrypted | Payload is wrapped in an encrypted envelope |
+
+##### Envelope header 
+| Field Size (in bytes) | Encoding  | Interpretation                              |
+|-----------------------|-----------|---------------------------------------------|
+|                     4 |    Uint32 | Key ID Length (Sk)                          |
+|                    Sk |    []Byte | Key ID Length                               |
+
+The key ID is an arbitrary string used to identify which key was used to encrypt the slice Date Encryption Key. Tink uses the notion of a "Key URI" across its various keystore integrations, so that the same mechanism can be used to identify a key regardless of how it is stored. We could use a simple scheme like `teleport://$KEY_ID_FROM_CONFIG_FILE`.
+
+#### Envelope Encryption Implementation
+
+NOTE: for the most part, this is an implementation detail. I'm assuming we will make a decision and delete this section before this RFD is adopted. In terms of implementation risk, I have a working proof of concept implemented with Tink.
+
+There are two competing options for implementing the envelope encryption:
+
+ * Defining our own envelope & key management system, or
+ * Using [Google Tink](https://developers.google.com/tink)
+
+##### Defining our own
+- **Pro:** Can be implemented using the tools in the standard `crypto` and `crypto/cipher` packages
+- **Pro:** Easier to verify for FIPS compliance as we don't have to wade through 3rd party code
+- **Con:** Have to write our own envelope encrypter, envelope format de-/serialiser & keystore. 
+
+None of this is particularly *difficult* (and we can gain a ***lot*** of useful inspiration from Tink), but it is worth noting that it needs to be done.
+
+##### Tink
+ - **Pro:** Integrated key management solution, should we later want to use a cloud-based key manager (e.g. Cloud Key Manager, AWS KMS) to manage master keys
+ - **Pro:** Pre-defined & implemented envelope format (uses protobuf under the hood, adds ~130B per slice)
+ - **Pro:** Plumbing between envelope decrypter and keystore is provided
+ - **Pro:** More intuitive API than `crypto` & `crypto/cipher`
+ - **Con:** Managing transitive third-party dependencies. Getting the PoC running was a challenge due to differing updated GRPC 
+ - **Con:** Does not support a locally-provided key out of the box. We still have to write our own keystore integration.
+ - **Con:** Harder to assert FIPS compliance. Unknown how it will react to `boringcrypto` builds.
```

Comment:
Should be something we can quickly test? I'd prefer to avoid building our own envelope format for sure

Reply:
@tcsc Sent you a doc on how I verified this last time around.

Reply:
I've verified that
1. Tink-enabled `teleport` builds with the amd64/fips build container
2. The boring crypto symbols appear in the unstripped `teleport` binary (as per @russjones' doc above)
3. Recordings are encrypted in the boringcrypto build of my proof-of-concept `teleport`
4. The Tink crypto primitive API defers to the stdlib `crypto/`packages for implementation (via manual audit)
That's probably enough for now.